// Copyright 2019 Istio Authors
//
//   Licensed under the Apache License, Version 2.0 (the "License");
//   you may not use this file except in compliance with the License.
//   You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.

syntax = "proto3";

package istio.context;

option cc_enable_arenas = true;

import "google/protobuf/wrappers.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/duration.proto";

import "google/api/expr/v1alpha1/syntax.proto";

message FilterConfig {
  google.api.expr.v1alpha1.Expr expr = 1;
}

message Request {
  map<string, string> headers = 1;
  google.protobuf.StringValue id = 2;
  google.protobuf.StringValue host = 3;
  google.protobuf.StringValue method = 4;
  google.protobuf.StringValue path = 5;
  google.protobuf.StringValue url_path = 6;
  map<string, string> query_params = 7;
  google.protobuf.StringValue reason = 8;
  google.protobuf.StringValue referer = 9;
  google.protobuf.StringValue scheme = 10;
  google.protobuf.Int64Value total_size = 11;
  google.protobuf.Int64Value size = 12;
  google.protobuf.Timestamp time = 13;
  google.protobuf.StringValue useragent = 14;
  google.protobuf.StringValue api_key = 15;
}

message RequestAuth {
  google.protobuf.StringValue principal = 1;
  google.protobuf.StringValue audiences = 2;
  google.protobuf.StringValue presenter = 3;
  map<string, string> claims = 4;
  google.protobuf.StringValue raw_claims = 5;
}

message Response {
  google.protobuf.Int64Value code = 1;
  google.protobuf.Duration duration = 2;
  map<string, string> headers = 3;
  google.protobuf.Int64Value total_size = 4;
  google.protobuf.Int64Value size = 5;
  google.protobuf.Timestamp time = 6;
  google.protobuf.StringValue grpc_status = 7;
  google.protobuf.StringValue grpc_message = 8;
}

message Connection {
  google.protobuf.StringValue event = 1;
  google.protobuf.StringValue id = 2;
  ConnectionReceived received = 3;
  ConnectionSent sent = 4;
  google.protobuf.Duration duration = 5;
  google.protobuf.BoolValue mtls = 6;
  google.protobuf.StringValue requested_server_name = 7;
}

message ConnectionReceived {
  google.protobuf.Int64Value bytes = 1;
  google.protobuf.Int64Value bytes_total = 2;
}

message ConnectionSent {
  google.protobuf.Int64Value bytes = 1;
  google.protobuf.Int64Value bytes_total = 2;
}

message Context {
  google.protobuf.StringValue protocol = 1;
  google.protobuf.StringValue proxy_error_code = 2;
  google.protobuf.Timestamp timestamp = 3;
  google.protobuf.Timestamp time = 4;
  ContextReporter reporter = 5;
}

message ContextReporter {
  google.protobuf.StringValue kind = 1;
  google.protobuf.StringValue uid = 2;
}

message Api {
  google.protobuf.StringValue service = 1;
  google.protobuf.StringValue version = 2;
  google.protobuf.StringValue operation = 3;
  google.protobuf.StringValue protocol = 4;
}

message Origin {
  google.protobuf.BytesValue ip = 1;
  google.protobuf.StringValue uid = 2;
  google.protobuf.StringValue user = 3;
}

/*
message Source {
  google.protobuf.StringValue uid = 1;
  google.protobuf.StringValue principal = 2;
}

message Destination {
  google.protobuf.StringValue uid = 1;
  google.protobuf.StringValue principal = 2;
  google.protobuf.Int64Value port = 3;
}

message Rbac {
  RbacPermissive permissive = 1;
}

message RbacPermissive {
  google.protobuf.StringValue response_code = 1;
  google.protobuf.StringValue effective_policy_id = 2;
}

message Check {
  google.protobuf.BoolValue cache_hit = 1;
  google.protobuf.Int64Value error_code = 2;
  google.protobuf.StringValue error_message = 3;
}

message Quota {
  google.protobuf.BoolValue cache_hit = 1;
}
*/
